feat: evm-swap example (#330)
LIT-Protocol/docs: docs/additional-examples/evm-swap.md

> # Cross Chain EVM Swap

Reviewer: "Performing a Cross-Chain Swap with PKPs and Lit Actions"

In this section, we will demonstrate how you can use PKPs and Lit Actions to perform a cross-chain swap. There are three parts to this flow:

> - Understanding Lit PKPs as escrow

Reviewer: > Understanding Lit PKPs as escrow

Consider "Using [Lit PKPs](https://developer.litprotocol.com/user-wallets/pkps/overview) as escrow accounts"

> - Lit Action

Reviewer: > Lit Action

can we be more specific here, like we were with the PKP usage? What are Lit Actions being used for in this flow? Let's highlight that

Reviewer: Maybe something like "using Lit Actions to define the swap parameters" ?

> - Implementing Mint/Grant/Burn with SDK

Reviewer: m/g/b is currently broken, so this example doesn't work right now? Or was their a workaround you used?

Author: it's a workaround, so what m/g/b does is the following three steps
- mints you a pkp 
- adds lit action as a permitted action
- burns the pkp

usual m/g/b function on smart contracts should be doing these in a single transaction (as a single functional call to smart contract) whereas in this example I've individually called these through SDK and later instead of burning the pkp, transferred the pkp to itself in a single javascript function

This simulates a m/g/b scenario, any pkp minted with the same flow through js functions, works exact the same as minted by m/g/b function on contracts

the difference which makes the network failing is that m/g/b function burns the pkp nft (sending it to 0x00 address) whereas I'm sending it to pkp's address, in both the ways, pkp acts like escrow whereas network fails on reading pkp owner as 0x00

Reviewer: gotcha, makes sense! Maybe we can add a callout box that makes this clear until the actual function is fixed

Author: sure, i'll add it

Author: does this look better?
https://github.com/LIT-Protocol/docs/pull/330/commits/30185e8fdd79532766306cf2736f33c5726e9f50#diff-0e36021eb2e9b13aa34e50ec77474d75d50bab8d18cba35c954841b81d9efee7R327

> ## 1) Understanding Lit PKPs as escrow

Reviewer: > Understanding Lit PKPs as escrow

"Using Lit PKPs as escrow accounts"

Author: ig it should be the same on lines 4 and 8, should I make it escrow accounts at both places or end it at escrow?

Reviewer: i think escrow is fine

> A Private key is meant for signing transactions which are executed on the blockchain, a network verifies if the transaction is a legitimate one and then executes it. With Lit Network we can generate a special type of private keys which can be programmed to sign transactions only when certain conditions are met. With this, we can create several escrow mechanisms, and we will look into how we can execute a cross-chain swap.

Reviewer: i think we can abstract away some of these details here, they feel a bit excessive. I think we can just briefly introduce PKPs (like we do here https://developer.litprotocol.com/user-wallets/pkps/overview in the overview) and then say that for this use case we are going to use them as escrow accounts to support the cross chain swap flow

Author: This was meant to give an overview of how we can "generally" use pkps for escrow-like cases and not only limit to defining pkps. As for bridging, messaging and swapping, the only thing that'll change will be the lit action.

I tried to make it use friendly but if it feels overwhelming, I can reduce it.

Reviewer: yea imo i think we can simplify things a bit. I think we should start by briefly introducing PKPs generally before honing in the specific use case of escrow

Author: check below
https://github.com/LIT-Protocol/docs/pull/330/commits/30185e8fdd79532766306cf2736f33c5726e9f50#diff-0e36021eb2e9b13aa34e50ec77474d75d50bab8d18cba35c954841b81d9efee7R10

> ### A swap example would be:

Reviewer: Let's just title this "Example"

> Suppose Alice holds 100 token A on Base and Bob holds 100 token B on Chronicles and they both agree to a swap of,

Reviewer: here can we just reference the actual tokens used in the example code? Like, "in the example below, Alice holds x tokens on Base and Bob holds x tokens on Chronicle. They agree to the following swap parameters:"

Author: I wrote this on testnets and so using my own custom-minted tokens. 
I designed it in a way if anyone wants to use the example as well, they can mint these tokens for themselves.

I can replace these with the token addresses (but that would make it too long, it's used A and B at many places), or for this example, we can assume eth and dai from the start and add a little disclaimer around conditions section where addresses are used that these are our representation of eth and dai on testnets.

Author: like i've added more context on revamping the below statement

Below are the parameters for the cross-chain swap. In the example below, we'll be transferring 4 tokens A on Sepolia for 8 tokens B on Chronicle Yellowstone. These tokens are minted by a custom contract.

Reviewer: cool, this works

- 4 token A from Alice to Bob on Base
- 8 token B from Bob to Alice on Chronicles

> We have an [example code](https://github.com/anshss/lit-evm-swap) for the below solution.

Reviewer: prob link the full example at the very bottom of this doc?

Author: I've added a link to the specific file at the end as well. But for a developer who is trying to implement this, I think it would be easier to see the codebase at the start. 

Lemme know if this doesn't look good and I'll remove this one from here.

Reviewer: sure, that works!

### How does this work?
When you generate a PKP on the LIT Network, a PKP NFT is first minted to your wallet, representing your ownership over the underlying private key residing on the LIT Network. This PKP NFT is responsible for managing auth methods over the private key.

Lit Actions are programmable code written in Javascript which can run on the LIT Network. We can program a Lit Action so that if certain permissions are met for a pkp, it must sign specific transactions that can be sent to the blockchain. We need to grant a Lit Action as a permitted action to the PKP to allow the Lit Action to generate signatures with it.

> 

Reviewer: > When you generate a PKP on the LIT Network, a PKP NFT is first minted to your wallet, representing your ownership over the underlying private key residing on the LIT Network. This PKP NFT is responsible for managing auth methods over the private key.

Lit Actions are programmable code written in Javascript which can run on the LIT Network. We can program a Lit Action so that if certain permissions are met for a pkp, it must sign specific transactions that can be sent to the blockchain. We need to grant a Lit Action as a permitted action to the PKP to allow the Lit Action to generate signatures with it.

These are all details we can abstract away from this guide imo. We can link to the intro pages for pkps and actions if folks are interested in learning about how they work generally. For the purpose of this guide, i think we should focus on the specific use of these in the implementation itself. So starting with "we first mint a PKP and then assign a Lit Action to it using the m/g/b method [link to code snippet]. Next, we ..." Going through the steps sequentially

Author: I've replaced this section with 
https://github.com/LIT-Protocol/docs/pull/330/commits/30185e8fdd79532766306cf2736f33c5726e9f50#diff-0e36021eb2e9b13aa34e50ec77474d75d50bab8d18cba35c954841b81d9efee7R19

> We can create an escrow pkp in the following way (Mint/ Grant/ Burn)

Reviewer: > We can create an escrow pkp in the following way (Mint/ Grant/ Burn)

Consider, "We can create a PKP "escrow" using the Mint/Grant/Burn [link to docs] method in the following way:"

Author: with [link to docs] you mean the link to current page where the example resides or exactly what page of the docs?

Reviewer: i was referring to the docs on m/g/b, but i realize that we don't have dedicated docs for this rn

Author: lmk, i can shoot another pr for that as well

- Mint a PKP on Lit

> - Grant a permitted Lit Action

Reviewer: "Grant a permitted Lit Action to sign with it"

> - Burn the PKP NFT so that no more auth methods can be modified

Reviewer: "Burn the PKP so that it can't be used to sign anything else other than what is outlined in the Lit Action logic"

> This approach will allow us to create scenarios where a private key can generate signs based on conditions and unlock robust opportunities like cross-chain swapping, messaging or bridging.

Reviewer: consider cutting this sentence, excess

> For a Swap Lit Action, both parties need to deposit their funds to the PKP and based on the conditions the PKP should be completing the swap:

Reviewer: I think maybe we should move these details to the 'lit action' section below? It seems more relevant to that

Author: moved

- Alice (or Bob) creates a Mint/Grant/Burn PKP
- Alice deposits 4 token A on Base Sepolia to the PKP

> - Bob deposits 8 token B on Chronicles to the PKP

Reviewer: "Chronicle" should be singular, not plural

Author: fixed this everywhere else as well.

- Alice (or Bob) executes the Lit Action.

> Lit Action execution, checks conditions and executes the swap from the PKP

Reviewer: > Lit Action execution, checks conditions and executes the swap from the PKP

"When the Lit Action is executed, each Lit node verifies that the conditions are met before provisioning their key share and executing the swap using the PKP"

> If the swap conditions aren't met then the Lit Action execution shall revert the funds to their depositors.

Reviewer: > If the swap conditions aren't met then the Lit Action execution shall revert the funds to their depositors.

"If the swap conditions aren't met, the Lit Action execution will revert and the tokens will be returned to the initial depositors (Alice and Bob)"

> ### Architecture for Mint/Grant/Burn

Reviewer: I think we should title this "Complete Swap Architecture"

Author: The diagram depicts how the general m/g/b function and lit actions built through it works. This should be the general flow of any such scenario as swap, messaging, bridging etc.

Swap definition is provided in the lit action

![accessControl](/img/mgb-flow.png)

## 2) Lit Action

> Now we'll dive into the depths of the Lit Action and explore how we can conduct conditional signing.

Reviewer: Let's be more specific. Something like "now we'll write the Lit Action that will be used to set the specific swap parameters that will be followed"

Author: Lit action is not exactly "setting the parameter" here. 
The same goes for changes in line 5.

A lit action is defining the code which remains immutable and works similarly to smart contracts in this scenario. We are defining swap conditions and what to do in those conditions.

Should we say this as "conditions that will be followed" or something else? I'm not able to come up with something suiting atm, or should we go with the parameter thing?

Reviewer: sure, 'conditions' works instead of 'parameters'

> Below are the parameters that decide the swapping conditions.

Reviewer: Can we provide more context on the example code? Something like "Below are the parameters for the cross-chain swap. In the example below, we'll be transferring 4 test ETH tokens on Sepolia for 8 (tstLPX?) tokens on Chronicle Yellowstone"

```js
const chainAParams = {
    from: "0x48e6a467852Fa29710AaaCDB275F85db4Fa420eB",
    to: "0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C",
    tokenAddress: "0xad50f302a957C165d865eD398fC3ca5A5A2cDA85",
    chain: "baseSepolia",
    amount: "4",
    decimals: 18,
    chainId: 84532,
};

const chainBParams = {
    from: "0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C",
    to: "0x48e6a467852Fa29710AaaCDB275F85db4Fa420eB",
    tokenAddress: "0x2dcA1a80c89c81C37efa7b401e2d20c1ED99C72F",
    chain: "yellowstone",
    amount: "8",
    decimals: 18,
    chainId: 175188,
};
/* 
deposit1: wallet A deposits on chain B, if action executes, funds are transferred to wallet B
deposit2: wallet B deposits on chain A, if action executes, funds are transferred to wallet A
*/
```
:::info

> On the chains section, we can specify any of the chains [supported](../resources/supported-chains) by LIT.

Reviewer: "You can use any of the chains [supported by Lit](https://developer.litprotocol.com/resources/supported-chains#programmable-key-pairs) when setting your own swap parameters"

:::

> You can find a generator which takes the above params to generate a Lit Action for swap: [swapActionGenerator.js](https://github.com/anshss/lit-evm-swap/blob/main/lit/swapActionGenerator.js)

Reviewer: > You can find a generator which takes the above params to generate a Lit Action for swap

"The following code can be used to generate the swap Lit Action based on the parameters above:"

> Remember, once a Mint/Grant/Burn PKP is created, anyone can execute the Lit Action on the Lit Network. This execution needs to define the parameters needed by the Lit Action to run. While generating a sign or checking conditions on the action, we need information around the pkp for which we are doing it. Due to this, we need to construct our Lit Action in a way that works with any PKP.

Reviewer: > Remember, once a Mint/Grant/Burn PKP is created, anyone can execute the Lit Action on the Lit Network. This execution needs to define the parameters needed by the Lit Action to run. While generating a sign or checking conditions on the action, we need information around the pkp for which we are doing it. Due to this, we need to construct our Lit Action in a way that works with any PKP.

Do we need to include this info here? Also, anyone could technically 'call' it, but wouldn't it depend on where the publisher stored it in the first place?

Author: for a m/g/b function on the [contracts](https://github.com/LIT-Protocol/LitNodeContracts/blob/main/contracts/lit-node/PKPNFT/PKPNFTFacet.sol#L232) or while simulating its approach like this example, we need to pass the cid for action in both the ways which makes a lit action an immutable piece of code. This makes it independent of the storage provider as anyway on using ipfs or arweave, your lit action would be immutable and visible to all.

As a specific session signature is not needed to just run a lit action on lit network, we need to construct our lit action in a way that it doesn't hardcode or depend on a specific pkp (in order, need to first construct the lit action and will then mint a pkp). This works like a wallet being created, then you fund the wallet, run the only action that can make a wallet create a sign, and get transactions as per conditions. 

The above info describes the nature of the action where an action is independent of the pkp data and made to run for a specific pkp. We can definitely short this information.

Reviewer 2: We _could_ make it to where a valid Auth Sig from either Alice or Bob has to be provided in order for the LA to execute, but not sure if it's worth adding this complexity to this example

Either way, I find the paragraph a bit confusing

Our Lit Action primarily focuses on three factors:

> - Access Conditions for checking funds on both chains

Reviewer: are these access control conditions?

Author: yes, renaming them to "access control conditions"

- Transaction objects for transferring funds between swap parties

> - Clawback Transaction objects for revert transferring funds to their owners

Reviewer: > Clawback Transaction objects for revert transferring funds to their owners

"Clawback transaction objects to revert transferred funds back to their owners if the swap parameters aren't met"

- Lit Action conditions which decide which transaction object to sign and return

> We'll use [Lit's Access Control](../sdk/access-control/evm/basic-examples#must-posess-at-least-one-erc20-token) here to verify if the conditions are being met for the swap. There need to be 2 conditions for a swap to execute,

Reviewer: "We'll use an Access Control Condition (ACC) here to verify if the swap conditions are met before it is executed. These conditions are:"

> - Have token A funds reached on Chain A?

Reviewer: isn't it checking if token a -> chain b and token b -> chain a?

Reviewer: or is it just verifying that both parties have the necessary balances to compete the swap in the first place?

Author: both the parties first deposit to the pkp, and a lit action is made to run, if the conditions are met the pkp forwards the funds to parties or reverts back to depositors.

chains remain constant in reference to the token, only the wallet holdings changes, an example would be,

wallet X and Y agree for a swap,
deposit1: wallet X deposits tokens B on chain B, if action executes, funds are transferred to wallet Y
deposit2: wallet Y deposits token A on chain A, if action executes, funds are transferred to wallet X

- Have token B funds reached on Chain B?

If yes, then forward the funds and complete the swap or else revert them.

A condition on Base Sepolia for checking if the PKP contains the exact amount of token A in Wei will look like this:

```js
const chainACondition = {
    conditionType: "evmBasic",
    contractAddress: "0xad50f302a957C165d865eD398fC3ca5A5A2cDA85",
    standardContractType: "ERC20",
    chain: "baseSepolia",
    method: "balanceOf",
    parameters: ["address"],
    returnValueTest: { comparator: "==", value: "4000000000000000000" },
};
// parameter field would be later replaced by the pkpAddress in the action
```

> There need to be two transaction objects, each of which transfers funds between the swap parties. Each of them will be calling function `transfer(address, uint256)` on the token's smart contract. This information is captured in the "data" field of the object. We can get that in the following way,

Reviewer: > There need to be two transaction objects, each of which transfers funds between the swap parties. Each of them will be calling function `transfer(address, uint256)` on the token's smart contract. This information is captured in the "data" field of the object. We can get that in the following way,

"We need to create two transaction objects, each of which will transfer the funds between the two swap parties that have been set. Each party will call the `transfer(address, uint256)` function on the token's smart contract. This information is captured in the "data" field of the object which we can fetch in the following way:"

```js
function generateCallData(counterParty, amount) {
    const transferInterface = new ethers.utils.Interface([
        "function transfer(address, uint256) returns (bool)",
    ]);
    return transferInterface.encodeFunctionData("transfer", [
        counterParty,
        amount,
    ]);
}
```

A transaction object would then look something like below

```js
let chainATransaction = {
    to: "0xad50f302a957C165d865eD398fC3ca5A5A2cDA85",
    gasLimit: "60000",
    from: "0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C",
    data: "0xa9059cbb000000000000000000000000291b0e3aa139b2bc9ebd92168575b5c6bad5236c0000000000000000000000000000000000000000000000003782dace9d900000",
    type: 2,
};
// "to" field specifies tokenA contract
// "from" field would be later replaced by the pkpAddress in the action
```

Clawback transaction objects are transactions that transfer funds back to the depositors if the swap conditions aren't met. These will have a different value for the `data` field and can be generated again using the `generateCallData()` function, with the `counterParty` value set to the depositor.

> Let's start with writing a Lit Action,

Reviewer: Maybe something like "Now that we've covered all of the relevant background information, let's write our Lit Action:"

    chainBClawbackTransaction = {...chainBClawbackTransaction, ...chainBGasConfig}
```

Here, we are substituting the `from` field in the transaction objects with `pkpAddress` and replacing the `parameters` the field in the conditions object with `pkpAddress`.

We then, extend our transaction object with the gas configs which will be received during the execution of the action. We'll look into this later.

```js
    const chainAConditionsPass = await Lit.Actions.checkConditions({
      conditions: [chainACondition],
      authSig: JSON.parse(authSig),
      chain: chainACondition.chain,
    });
  
    const chainBConditionsPass = await Lit.Actions.checkConditions({
      conditions: [chainBCondition],
      authSig: JSON.parse(authSig),
      chain: chainBCondition.chain,
    });

    console.log("chainAConditionsPass: ", chainAConditionsPass, "chainBConditionsPass: ", chainBConditionsPass);
```

Now our action checks if the conditions are passed on both chains, it takes `authSig` as a parameter here and then logs the result of both condition checks. We'll look into auth sigs while we learn to execute lit actions.

> For generating signatures inside our Lit Action, we'll use `Lit.Actions.signEcdsa`. This method allows us to use a pkp (which lets an action generate a sign) for generating signatures. Our Lit Action will now decide based on checked conditions and generate appropriately signed transactions which we can send to the blockchain later.

Reviewer: > For generating signatures inside our Lit Action, we'll use `Lit.Actions.signEcdsa`. This method allows us to use a pkp (which lets an action generate a sign) for generating signatures. Our Lit Action will now decide based on checked conditions and generate appropriately signed transactions which we can send to the blockchain later.

Consider "To generate signatures inside of our Lit Action, we'll use the `Lit.Actions.signEcdsa` method. Our Lit Action will verify that the conditions we've specified are met before generating the appropriate signed transactions which we'll be able to send to chain later"

        publicKey: pkpPublicKey,
        sigName: "chainBSignature",
      });
      Lit.Actions.setResponse({
        response: JSON.stringify({
          chainBTransaction: chainBClawbackTransaction,
        }),
      });
      return;
    }

    const generateSwapTransactions = async () => {
      await LitActions.signEcdsa({
        toSign: hashTransaction(chainATransaction),
        publicKey: pkpPublicKey,
        sigName: "chainASignature",
      });
      await LitActions.signEcdsa({
        toSign: hashTransaction(chainBTransaction),
        publicKey: pkpPublicKey,
        sigName: "chainBSignature",
      });
      Lit.Actions.setResponse({
        response: JSON.stringify({ chainATransaction, chainBTransaction }),
      });
    };
```

As everything on the blockchain works on hexadecimal hashes, we need to hash our transactions as well.

```js
    const hashTransaction = (tx) => {
      return ethers.utils.arrayify(
        ethers.utils.keccak256(
          ethers.utils.arrayify(ethers.utils.serializeTransaction(tx)),
        ),
      );
    }

    Lit.Actions.setResponse({ response: "Conditions for swap not met!" });
  }
go();
```

> A must `Lit.Actions.setResponse` needs to be set before we are closing the execution of a Lit Action. This is the response which is shown after the execution of a Lit Action.

Reviewer: > A must `Lit.Actions.setResponse` needs to be set before we are closing the execution of a Lit Action. This is the response which is shown after the execution of a Lit Action.

"We need to remember to set `Lit.Actions.setResponse` before we finish executing the Lit Action. This defines the response that will be shown when the Lit Action finishes executing"

> Complete Lit Action can be found at [swapAction.js](https://github.com/anshss/lit-evm-swap/blob/main/lit/swapAction.js).

Reviewer: "The complete swap Lit Action can be found [here](https://github.com/anshss/lit-evm-swap/blob/main/lit/swapAction.js)"

> ## 3) Implementing Mint/Grant/Burn with SDK

Reviewer: This goes back to my earlier question, but m/g/b is broken right now right?

Also, is the point of this section to go over the full flow start to finish?

Author: Our SDK does not support mint/grant/burn calls at this moment so we can keep this alternate approach for a time even after the fix.

The only change we'll later need to make would be in this [function](https://github.com/LIT-Protocol/docs/pull/330/commits/e1a8353388296a6bbdbda28fb4f81d9681c27d33#diff-0e36021eb2e9b13aa34e50ec77474d75d50bab8d18cba35c954841b81d9efee7R342).

> ### Installing Packages

Reviewer: "Installing the relevant packages"

```bash
npm i @lit-protocol/lit-node-client @lit-protocol/contracts-sdk 
@simplewebauthn/browser  @lit-protocol/types ethers
```

### Imports Declarations

```js
import { LitNodeClient } from "@lit-protocol/lit-node-client";
import { LitContracts } from "@lit-protocol/contracts-sdk";
import { LitNetwork, AuthMethodScope, LIT_CHAINS } from "@lit-protocol/constants";
import { LitAbility } from "@lit-protocol/types";
import {
    LitActionResource,
    createSiweMessageWithRecaps,
    generateAuthSig,
    LitPKPResource,
} from "@lit-protocol/auth-helpers";
import { ethers } from "ethers";
import bs58 from "bs58";
```

> ### Create Wallet instances

Reviewer: "Creating our two wallet instances"

These wallets will represent the two parties involved in the swap.

```js
async function getWalletA() {
    const provider = new ethers.providers.JsonRpcProvider(
        `https://yellowstone-rpc.litprotocol.com/`
    );
    const wallet = new ethers.Wallet(
        process.env.NEXT_PUBLIC_PRIVATE_KEY_1,
        provider
    );
    return wallet;
}

async function getWalletB() {
    const provider = new ethers.providers.JsonRpcProvider(
        `https://yellowstone-rpc.litprotocol.com/`
    );
    const wallet = new ethers.Wallet(
        process.env.NEXT_PUBLIC_PRIVATE_KEY_2,
        provider
    );
    return wallet;
}c
```

> ### Create Lit Action and upload it to IPFS

Reviewer: "Creating our Lit Action and uploading it to IPFS"

Once we are done with our Lit Action, we need to upload it to IPFS so we get an immutable CID which will always point to our action.

```js
export async function createLitAction() {
    console.log("creating lit action..");
    const action = createERC20SwapLitAction(chainAParams, chainBParams);
    const ipfsCid = await uploadViaPinata(action);

    console.log("Lit Action code:\n", action);
    console.log("IPFS CID: ", ipfsCid);
    return ipfsCid;
}
```

### Create a Mint/Grant/Burn PKP
We'll create a single function to perform 3 calls at once

- Mint a PKP with the user's wallet
- Add the Lit Action as a permitted auth Method

> - Burn the PKP NFT by transferring it to itself so auth methods become immutable

Reviewer: > Burn the PKP NFT by transferring it to itself so auth methods become immutable

I think the important thing to emphasize here is that this method ensures that the PKP can only be used to sign based on the initial conditions that you define and nothing else

Reviewer: which you do a good job of in the sentence below

Author: I can replace it to make it the same as the line 28
"Burn the PKP so that it can't be used to sign anything else other than what is outlined in the Lit Action logic"

The owner of pkp nft can send txn to smart contracts and modify auth methods so we burn it.
The primary reason for burning a pkp nft is to prevent any further modifications to the auth methods, even though the ultimate goal aligns with the above statement. I wrote this way to give more clarity to the developer on the thought process.

If it feels like not to be needed, I can replace it.

Reviewer: this sounds good to me

    console.log("minting started..");
    const signerA = await getWalletA();

// instantiating Contracts SDK
    const litContracts = new LitContracts({
        signer: signerA,
        network: LitNetwork.DatilDev,
        debug: false,
    });
    await litContracts.connect();

// minting a new PKP
    const mintPkp = await litContracts.pkpNftContractUtils.write.mint();
    const pkp = mintPkp.pkp;
    console.log("PKP: ", pkp);

    console.log("adding permitted action..");

// adding a permitted lit action
    await litContracts.addPermittedAction({
        pkpTokenId: pkp.tokenId,
        ipfsId: action_ipfs,
        authMethodScopes: [AuthMethodScope.SignAnything],
    });

    console.log("transfer started..");

// burning the PKP NFT
    const transferPkpOwnership =
        await litContracts.pkpNftContract.write.transferFrom(
            signerA.address,
            pkp.ethAddress,
            pkp.tokenId,
            {
                gasLimit: 125_000,
            }
        );

    const receipt = await transferPkpOwnership.wait();

    console.log(
        "Transferred PKP ownership to itself: ",
        receipt
    );
    return pkp;
}
```

> ### Deposit Methods

Reviewer: "Depositing tokens for the swap"

            results.response.chainATransaction,
            signatureA
        )
    );
    console.log(tx1);
    
    const receipt1 = await tx1.wait();
    const blockExplorer1 = LIT_CHAINS[chainAParams.chain].blockExplorerUrls[0];
    
    console.log(`tx: ${blockExplorer1}/tx/${receipt1.transactionHash}`);
}

async function executeTxB(results, chainBProvider) {
    const signatureB = formatSignature(results.signatures.chainBSignature);
    const tx2 = await chainBProvider.sendTransaction(
        ethers.utils.serializeTransaction(
            results.response.chainBTransaction,
            signatureB
        )
    );
    const receipt2 = await tx2.wait();
    const blockExplorer2 = LIT_CHAINS[chainBParams.chain].blockExplorerUrls[0];

    console.log(`tx: ${blockExplorer2}/tx/${receipt2.transactionHash}`);
}
```

A signed transaction returned by the Lit Network has a different structure for signatures that we need to reconstruct as follows:

```js
function formatSignature(signature) {
    const encodedSig = ethers.utils.joinSignature({
        v: signature.recid,
        r: `0x${signature.r}`,
        s: `0x${signature.s}`,
    });
    return encodedSig;
}

> ```

Reviewer: I think it could be worth including a "Summary" section which recaps all of the above steps and defines success criteria

Author: https://github.com/LIT-Protocol/docs/pull/330/commits/30185e8fdd79532766306cf2736f33c5726e9f50#diff-0e36021eb2e9b13aa34e50ec77474d75d50bab8d18cba35c954841b81d9efee7R656

```js
const chainACondition = {
    conditionType: "evmBasic",
    contractAddress: "0xad50f302a957C165d865eD398fC3ca5A5A2cDA85",
    standardContractType: "ERC20",
    chain: "baseSepolia",
    method: "balanceOf",
    parameters: ["address"],
    returnValueTest: { comparator: "==", value: "4000000000000000000" },
};
// parameter field would be later replaced by the pkpAddress in the action
```
We need to create two transaction objects, each of which will transfer the funds between the two swap parties that have been set. Each party will call the `transfer(address, uint256)` function on the token's smart contract. This information is captured in the "data" field of the object which we can fetch in the following way:

```js
function generateCallData(counterParty, amount) {
    const transferInterface = new ethers.utils.Interface([
        "function transfer(address, uint256) returns (bool)",
    ]);
    return transferInterface.encodeFunctionData("transfer", [
        counterParty,
        amount,
    ]);
}
```

A transaction object would then look something like below

```js
let chainATransaction = {
    to: "0xad50f302a957C165d865eD398fC3ca5A5A2cDA85",
    gasLimit: "60000",

>     from: "0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C",

Reviewer 2: I think you could just omit this property. We don't usually specify the `from` property in an ETH transaction since it's derived from the signature

    let chainATransaction = {"to":"0xad50f302a957C165d865eD398fC3ca5A5A2cDA85","gasLimit":"60000","from":"0x48e6a467852Fa29710AaaCDB275F85db4Fa420eB","data":"0xa9059cbb000000000000000000000000291b0e3aa139b2bc9ebd92168575b5c6bad5236c0000000000000000000000000000000000000000000000003782dace9d900000","type":2}
    let chainBTransaction = {"to":"0x2dcA1a80c89c81C37efa7b401e2d20c1ED99C72F","gasLimit":"60000","from":"0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C","data":"0xa9059cbb00000000000000000000000048e6a467852fa29710aaacdb275f85db4fa420eb0000000000000000000000000000000000000000000000006f05b59d3b200000","type":2}
    let chainAClawbackTransaction = {"to":"0xad50f302a957C165d865eD398fC3ca5A5A2cDA85","gasLimit":"60000","from":"0x48e6a467852Fa29710AaaCDB275F85db4Fa420eB","data":"0xa9059cbb00000000000000000000000048e6a467852fa29710aaacdb275f85db4fa420eb0000000000000000000000000000000000000000000000003782dace9d900000","type":2}

>     let chainBClawbackTransaction = {"to":"0x2dcA1a80c89c81C37efa7b401e2d20c1ED99C72F","gasLimit":"60000","from":"0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C","data":"0xa9059cbb000000000000000000000000291b0e3aa139b2bc9ebd92168575b5c6bad5236c0000000000000000000000000000000000000000000000006f05b59d3b200000","type":2}

Reviewer 2: Wouldn't we want to pass these in as a parameter to the LA? This way the LA could be reused by anyone for swaps

Author: Passing these as a parameter to the la would disrupt the meaning of "trustless swap" being offered by lit protocol as the action could then be ran with wrong values at runtime as well

I've made a separate la generator script where you pass a simple set of swap parameters and it will return you a swap la, have mentioned that in the docs
my flow is to pass simple set of swap parameter like 
```
const chainAParams = {
    from: "0x48e6a467852Fa29710AaaCDB275F85db4Fa420eB",
    to: "0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C",
    tokenAddress: "0xad50f302a957C165d865eD398fC3ca5A5A2cDA85",
    chain: "baseSepolia",
    amount: "4",
    decimals: 18,
    chainId: 84532,
};

const chainBParams = {
    from: "0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C",
    to: "0x48e6a467852Fa29710AaaCDB275F85db4Fa420eB",
    tokenAddress: "0x2dcA1a80c89c81C37efa7b401e2d20c1ED99C72F",
    chain: "yellowstone",
    amount: "8",
    decimals: 18,
    chainId: 175188,
};
```
to swapGenerator, then that cook you a la

>     chainATransaction.from = chainBTransaction.from = pkpAddress;

Reviewer 2: Don't think this is necessary either?

          chainBTransaction: chainBClawbackTransaction,
        }),
      });
      return;
    }

    const generateSwapTransactions = async () => {
      await LitActions.signEcdsa({
        toSign: hashTransaction(chainATransaction),
        publicKey: pkpPublicKey,
        sigName: "chainASignature",
      });
      await LitActions.signEcdsa({
        toSign: hashTransaction(chainBTransaction),
        publicKey: pkpPublicKey,
        sigName: "chainBSignature",
      });
      Lit.Actions.setResponse({
        response: JSON.stringify({ chainATransaction, chainBTransaction }),
      });
    };
```

As everything on the blockchain works on hexadecimal hashes, we need to hash our transactions as well.

```js
    const hashTransaction = (tx) => {
      return ethers.utils.arrayify(
        ethers.utils.keccak256(
          ethers.utils.arrayify(ethers.utils.serializeTransaction(tx)),
        ),
      );
    }

    Lit.Actions.setResponse({ response: "Conditions for swap not met!" });
  }
go();
```

We need to remember to set `Lit.Actions.setResponse` before we finish executing the Lit Action. This defines the response that will be shown when the Lit Action finishes executing.

> The complete swap Lit Action can be found [here](https://github.com/anshss/lit-evm-swap/blob/main/lit/swapAction.js).

Reviewer 2: ```suggestion
The complete swap Lit Action can be found [here](https://github.com/LIT-Protocol/evm-swap-example/blob/main/lit/litAction-example.js).
```

## 3) Implementing Mint/Grant/Burn with SDK

### Installing and importing the relevant packages

```bash

> npm i @lit-protocol/lit-node-client @lit-protocol/contracts-sdk 
> @simplewebauthn/browser  @lit-protocol/types ethers

Reviewer 2: ```suggestion
npm i \
@lit-protocol/lit-node-client \
@lit-protocol/contracts-sdk \
@simplewebauthn/browser \
@lit-protocol/types ethers \
```

```

```js
import { LitNodeClient } from "@lit-protocol/lit-node-client";
import { LitContracts } from "@lit-protocol/contracts-sdk";

> import { LitNetwork, AuthMethodScope, LIT_CHAINS } from "@lit-protocol/constants";

Reviewer 2: ```suggestion
import { LitNetwork, AuthMethodScope, LIT_RPC, LIT_CHAINS } from "@lit-protocol/constants";
```

import { LitAbility } from "@lit-protocol/types";
import {
    LitActionResource,
    createSiweMessageWithRecaps,
    generateAuthSig,
    LitPKPResource,
} from "@lit-protocol/auth-helpers";
import { ethers } from "ethers";
import bs58 from "bs58";
```

### Creating our two wallet instances

These wallets will represent the two parties involved in the swap.

```js
async function getWalletA() {
    const provider = new ethers.providers.JsonRpcProvider(

>         `https://yellowstone-rpc.litprotocol.com/`

Reviewer 2: ```suggestion
        LIT_RPC.CHRONICLE_YELLOWSTONE
```

    );
    const wallet = new ethers.Wallet(
        process.env.NEXT_PUBLIC_PRIVATE_KEY_1,
        provider
    );
    return wallet;
}

async function getWalletB() {
    const provider = new ethers.providers.JsonRpcProvider(

>         `https://yellowstone-rpc.litprotocol.com/`

Reviewer 2: ```suggestion
        LIT_RPC.CHRONICLE_YELLOWSTONE
```

:::

We'll be using `mintNextAndAddAuthMethods` to perform 3 calls in a single transaction:

- Mint a PKP with the user's wallet
- Add the Lit Action as a permitted auth method
- 'Burn' the PKP NFT by transferring it to itself so that no additional auth methods can be added. Effectively, this offers the same functionality as the M/G/B method

Now our PKP can only be used to execute the Lit Action we permitted when we added it as an auth method. It will only generate signatures when the Lit Action is running and the conditions we set have been met.

```js
export async function mintGrantBurnPKP(action_ipfs, mintedPKP) {
    console.log("minting started..");
    const signerA = await getWalletA();

    const litContracts = new LitContracts({
        signer: signerA,
        network: LitNetwork.DatilDev,
        debug: false,
    });

    const bytesAction = await stringToBytes(action_ipfs);
    const pkpMintCost = await litContracts.pkpNftContract.read.mintCost();

    await litContracts.connect();

    const tx =
        await litContracts.pkpHelperContract.write.mintNextAndAddAuthMethods(

>             AuthMethodType.LitAction,
>             [AuthMethodType.LitAction],
>             [bytesAction],
>             ["0x"],
>             [[AuthMethodScope.SignAnything]],
>             false,
>             true,

Reviewer 2: ```suggestion
            AuthMethodType.LitAction, // keyType
            [AuthMethodType.LitAction], // permittedAuthMethodTypes
            [bytesAction], // permittedAuthMethodIds
            ["0x"], // permittedAuthMethodPubkeys
            [[AuthMethodScope.SignAnything]], // permittedAuthMethodScopes
            false, // addPkpEthAddressAsPermittedAddress
            true, // sendPkpToItself
```

            {
                value: pkpMintCost,

>                 maxFeePerGas: ethers.BigNumber.from("1800000000"),

Reviewer 2: I don't think this property is required?

Author: my transaction was failing so I had to configure the gases manually

            results.response.chainATransaction,
            signatureA
        )
    );
    console.log(tx1);
    
    const receipt1 = await tx1.wait();
    const blockExplorer1 = LIT_CHAINS[chainAParams.chain].blockExplorerUrls[0];
    
    console.log(`tx: ${blockExplorer1}/tx/${receipt1.transactionHash}`);
}

async function executeTxB(results, chainBProvider) {
    const signatureB = formatSignature(results.signatures.chainBSignature);
    const tx2 = await chainBProvider.sendTransaction(
        ethers.utils.serializeTransaction(
            results.response.chainBTransaction,
            signatureB
        )
    );
    const receipt2 = await tx2.wait();
    const blockExplorer2 = LIT_CHAINS[chainBParams.chain].blockExplorerUrls[0];

    console.log(`tx: ${blockExplorer2}/tx/${receipt2.transactionHash}`);
}
```

A signed transaction returned by the Lit Network has a different structure for signatures that we need to reconstruct as follows:

```js
function formatSignature(signature) {
    const encodedSig = ethers.utils.joinSignature({
        v: signature.recid,
        r: `0x${signature.r}`,
        s: `0x${signature.s}`,
    });
    return encodedSig;
}
```

> You can also explore other methods, such as checking permissions on the PKP, checking the fund status of each wallet, or the PKP itself [here](https://github.com/anshss/lit-evm-swap/blob/main/lit/utils.js).

Reviewer 2: ```suggestion
You can also explore other methods, such as checking permissions on the PKP, checking the fund status of each wallet, or the PKP itself [here](https://github.com/LIT-Protocol/evm-swap-example/blob/main/lit/utils.js).
```

# Performing a Cross-Chain Swap with PKPs and Lit Actions
In this section, we will demonstrate how you can use PKPs and Lit Actions to perform a cross-chain swap. There are three parts to this flow:

- Using Lit PKPs as escrow
- Using Lit Actions to define swap conditions

> - Implementing Mint/Grant/Burn with SDK

Reviewer: > Implementing Mint/Grant/Burn with SDK

"Implementing Mint/Grant/Burn with the Lit SDK"

## 1) Using Lit PKPs as escrow

> [PKPs](../user-wallets/pkps/overview) are decentralized key pairs generated by the Lit Network. Using Lit Actions, we can program them to sign transactions under specific conditions, enabling escrow mechanisms. We’ll explore how to execute a cross-chain swap with this setup.

Reviewer: > [PKPs](../user-wallets/pkps/overview) are decentralized key pairs generated by the Lit Network. Using Lit Actions, we can program them to sign transactions under specific conditions, enabling escrow mechanisms. We’ll explore how to execute a cross-chain swap with this setup.

"[PKPs](../user-wallets/pkps/overview) are decentralized key pairs generated by the Lit Network. Using [Lit Actions](https://developer.litprotocol.com/sdk/serverless-signing/overview), we can program PKPs to sign transactions when our pre-defined conditions are met. In the following example, we'll use this functionality to execute a cross-chain swap"

### Example:
Suppose Alice holds 100 token A on Base and Bob holds 100 token B on Chronicle and they both agree to a swap of,
- 4 token A from Alice to Bob on Base
- 8 token B from Bob to Alice on Chronicle

> We have an [example code](https://github.com/anshss/lit-evm-swap) for the below solution.

Reviewer: > We have an [example code](https://github.com/anshss/lit-evm-swap) for the below solution.

"A complete code example can be found [here](https://github.com/anshss/lit-evm-swap)"

- However, let's first ensure that we move the referenced example to the Lit Protocol repo first (in developer guides?)

Author: yup, moved it

### How does this work?
1.	Mint a PKP: Start by minting a PKP on the Lit Network.

> 2.	Assign a Lit Action: Grant a specific Lit Action to the PKP using the Mint/Grant/Burn (M/G/B) approach, allowing it to generate signatures according to the programmed conditions.

Reviewer: > Assign a Lit Action: Grant a specific Lit Action to the PKP using the Mint/Grant/Burn (M/G/B) approach, allowing it to generate signatures according to the programmed conditions.

"Assign a Lit Action: Grant a specific Lit Action the ability to use your PKP using the Mint/Grant/Burn (M/G/B) method which will allow it to generate signatures according to the conditions you set."

> 3.	Burn the PKP: Finally, burn the PKP to ensure it can only be used within the confines of the Lit Action logic.

Reviewer: In step 2 you already specify that the pkp will be minted with m/g/b so this feels repetitive

Author: ig it should be necessary to describe the flow once in the start before moving forward or there wouldn't be much context around what the user is doing till he reaches step 3

later in [step 3](https://github.com/anshss/docs/blob/evmswap/docs/additional-examples/evm-swap.md?plain=1#L333) is a repetition to highlight the working of the function

lmk your view

> Learn more about Lit Actions [here](../sdk/serverless-signing/overview).

Reviewer: this should link to docs on the m/g/b method

Author: ig we do not have a section for that yet. 
can you please drop the link that needs to be pointed here?

> ### Architecture for Mint/Grant/Burn

Reviewer: since this diagram includes creating session sigs and the eventual execution of the lit action itself, i think this should be titled "Architecture Overview" or something to that effect

Reviewer 2: In this Diagram, what's being represented as `Home`?

Author: ig it represents the current app
@Ansonhkg crafted this, maybe he could tell more

![accessControl](/img/mgb-flow.png)

> ## 2) Lit Action to define swap conditions

Reviewer: "Creating a Lit Action to define your swap conditions"

> For a Swap Lit Action, both parties need to deposit their funds to the PKP and based on the conditions the PKP should be completing the swap:

Reviewer: > For a Swap Lit Action, both parties need to deposit their funds to the PKP and based on the conditions the PKP should be completing the swap:

"For our swap Lit Action, both parties will need to deposit their funds to the PKP 'escrow'. The PKP will complete the swap when the conditions we set in the Lit Action have been met"

const chainBParams = {
    from: "0x291B0E3aA139b2bC9Ebd92168575b5c6bAD5236C",
    to: "0x48e6a467852Fa29710AaaCDB275F85db4Fa420eB",
    tokenAddress: "0x2dcA1a80c89c81C37efa7b401e2d20c1ED99C72F",
    chain: "yellowstone",
    amount: "8",
    decimals: 18,
    chainId: 175188,
};
/* 
deposit1: wallet X deposits on chain B, if action executes, funds are transferred to wallet Y
deposit2: wallet Y deposits on chain A, if action executes, funds are transferred to wallet X
*/
```
:::info
You can use any of the chains [supported](../resources/supported-chains) by Lit when setting your own swap parameters
:::

The following code can be used to generate the swap Lit Action based on the parameters above: [swapActionGenerator.js](https://github.com/anshss/lit-evm-swap/blob/main/lit/swapActionGenerator.js)

Remember, once a Mint/Grant/Burn PKP is created, anyone can execute the Lit Action on the Lit Network. This execution needs to define the parameters needed by the Lit Action to run. While generating a sign or checking conditions on the action, we need information around the pkp for which we are doing it. Due to this, we need to construct our Lit Action in a way that works with any PKP.

> Our Lit Action primarily focuses on three factors:

Reviewer: *4 factors

- Access control conditions for checking funds on both chains

> - Transaction objects for transferring funds between swap parties

Reviewer: "Transaction objects for transferring funds between our two swap parties"

- Clawback transaction objects to revert transferred funds back to their owners if the swap parameters aren't met

> - Lit action conditions which decide which transaction object to sign and return

Reviewer: "[Lit Action conditions](https://developer.litprotocol.com/sdk/access-control/lit-action-conditions) to determine which transaction objects should be signed and returned"

These wallets will represent the two parties involved in the swap.

```js
async function getWalletA() {
    const provider = new ethers.providers.JsonRpcProvider(
        `https://yellowstone-rpc.litprotocol.com/`
    );
    const wallet = new ethers.Wallet(
        process.env.NEXT_PUBLIC_PRIVATE_KEY_1,
        provider
    );
    return wallet;
}

async function getWalletB() {
    const provider = new ethers.providers.JsonRpcProvider(
        `https://yellowstone-rpc.litprotocol.com/`
    );
    const wallet = new ethers.Wallet(
        process.env.NEXT_PUBLIC_PRIVATE_KEY_2,
        provider
    );
    return wallet;
}
```

### Creating our Lit Action and uploading it to IPFS

Once we are done with our Lit Action, we need to upload it to IPFS so we get an immutable CID which will always point to our action.

```js
export async function createLitAction() {
    console.log("creating lit action..");
    const action = createERC20SwapLitAction(chainAParams, chainBParams);
    const ipfsCid = await uploadViaPinata(action);

    console.log("Lit Action code:\n", action);
    console.log("IPFS CID: ", ipfsCid);
    return ipfsCid;
}
```

### Create a Mint/Grant/Burn PKP

:::info

> Currently, direct calls to the smart contract’s m/g/b function are currently failing with the Lit Network. We’ll replicate the same functionality using the SDK.

Reviewer: > Currently, direct calls to the smart contract’s m/g/b function are currently failing with the Lit Network. We’ll replicate the same functionality using the SDK.

"The M/G/B function is currently broken. For this example, we'll create a workaround the replicates the same functionality using the Lit SDK"

Reviewer 2: Why is this currently broken? Do we have an ETA for a fix?

Author: Not exactly, Adarsh has been working on the issue earlier. Check [LIT-3762](https://linear.app/litprotocol/issue/LIT-3762/cant-authenticate-for-burnt-pkp).

Reviewer 2: We could exclude this callout since using `mintNextAndAddAuthMethods` with `sendPkpToItself` enabled, and the only Auth Method being the LA is effectively the same thing as MGB

:::

> We'll create a single function to perform 3 calls at once

Reviewer: please add a ':' to the end of this sentence

- Mint a PKP with the user's wallet

> - Add the Lit Action as a permitted auth Method

Reviewer: 'method' should be lower case

> - Burn the PKP NFT by transferring it to itself so auth methods become immutable

Reviewer: > Burn the PKP NFT by transferring it to itself so auth methods become immutable

"'Burn' the PKP NFT by transferring it to itself so that no additional auth methods can be added. Effectively, this offers the same functionality as the M/G/B method"

> Now, our PKP will only be used to execute the Lit Action we permitted. It will generate signatures only when the Lit Action is running and not in any other situations. We have already designed the Lit Action to create the correct signed transactions based on the conditions.

Reviewer: > Now, our PKP will only be used to execute the Lit Action we permitted. It will generate signatures only when the Lit Action is running and not in any other situations. We have already designed the Lit Action to create the correct signed transactions based on the conditions.

"Now our PKP can only be used to execute the Lit Action we permitted when we added it as an auth method. It will only generate signatures when the Lit Action is running and the conditions we set have been met"

    console.log("minting started..");
    const signerA = await getWalletA();

// instantiating Contracts SDK
    const litContracts = new LitContracts({
        signer: signerA,
        network: LitNetwork.DatilDev,
        debug: false,
    });
    await litContracts.connect();

// minting a new PKP
    const mintPkp = await litContracts.pkpNftContractUtils.write.mint();
    const pkp = mintPkp.pkp;
    console.log("PKP: ", pkp);

    console.log("adding permitted action..");

// adding a permitted lit action
    await litContracts.addPermittedAction({
        pkpTokenId: pkp.tokenId,
        ipfsId: action_ipfs,
        authMethodScopes: [AuthMethodScope.SignAnything],
    });

    console.log("transfer started..");

// burning the PKP NFT
    const transferPkpOwnership =
        await litContracts.pkpNftContract.write.transferFrom(
            signerA.address,
            pkp.ethAddress,
            pkp.tokenId,
            {
                gasLimit: 125_000,
            }
        );

    const receipt = await transferPkpOwnership.wait();

    console.log(
        "Transferred PKP ownership to itself: ",
        receipt
    );
    return pkp;

> }

Reviewer 2: This can be done in a single transaction using `litContracts.pkpHelperContract.write.mintNextAndAddAuthMethods` see [here](https://github.com/LIT-Protocol/custom-auth-telegram-example/blob/7aee88a01bc9feb42dd81f3f1cbfb17b5d0f470f/src/mintPkp.ts#L55-L71)

Author: I've earlier done this for simplicity and to improve readiness but would update it to use `mintNextAndAddAuthMethods ` now

Author: ah this slipped, will fix it in both code and docs

    
    console.log(`tx: ${blockExplorer1}/tx/${receipt1.transactionHash}`);
}

async function executeTxB(results, chainBProvider) {
    const signatureB = formatSignature(results.signatures.chainBSignature);
    const tx2 = await chainBProvider.sendTransaction(
        ethers.utils.serializeTransaction(
            results.response.chainBTransaction,
            signatureB
        )
    );
    const receipt2 = await tx2.wait();
    const blockExplorer2 = LIT_CHAINS[chainBParams.chain].blockExplorerUrls[0];

    console.log(`tx: ${blockExplorer2}/tx/${receipt2.transactionHash}`);
}
```

A signed transaction returned by the Lit Network has a different structure for signatures that we need to reconstruct as follows:

```js
function formatSignature(signature) {
    const encodedSig = ethers.utils.joinSignature({
        v: signature.recid,
        r: `0x${signature.r}`,
        s: `0x${signature.s}`,
    });
    return encodedSig;
}
```

You can also explore other methods, such as checking permissions on the PKP, checking the fund status of each wallet, or the PKP itself [here](https://github.com/anshss/lit-evm-swap/blob/main/lit/utils.js).

> ## Summary

Reviewer: Before you list the bullet points, consider adding "In the following example you should have learned how to:"

> - Created transactions and clawback transactions objects based on swap parameters.

Reviewer: "Created transaction and clawback transaction objects based on specific swap parameters"

> - Wrote a Lit Action that returns signed transactions based on conditions.

Reviewer: "Wrote a Lit Action that returns signed blockchain transactions based on the specific conditions you defined"

- Minted a PKP and assigned it the Lit Action.
- Burned the PKP to restrict its signing ability to the Lit Action only.

> - Deposited funds to the PKP on both chains.

Reviewer: "Deposited funds to the PKP on two blockchains where our swap will be conducted between"

> - Executed the Lit Action on the Lit Network to receive signed transactions.

Reviewer: "Executed the Lit Action on the Lit network to return our signed transaction objects"
